pkg/apis/pipeline/v1beta1/result_validation.go

```diff
@@ -31,9 +31,5 @@ func (tr TaskResult) Validate(ctx context.Context) (errs *apis.FieldError) {
 		return errs.Also(ValidateEnabledAPIFields(ctx, "results type", config.AlphaAPIFields))
 	}
 
-	if tr.Type != ResultsTypeString {
-		return apis.ErrInvalidValue(tr.Type, "type", fmt.Sprintf("type must be string"))
-	}
```

Comment:
Oh, this is not what I was expecting, now we will allow for "type" to be anything?
In fact you had to delete this unit test, which is not something we would like to allow.

```yaml
		fields: fields{
			Steps: validSteps,
			Results: []v1beta1.TaskResult{{
				Name:        "MY-RESULT",
				Type:        "wrong",
				Description: "my great result",
			}},
		},
```

The only case we need to allow for is type not being set at all, which might be the case for resources created before type existed.

Reply:
Sorry I missed this, Andrea is right. I think we want something like:
```
if tr.Type == "" {
  return nil
}
if tr.Type != ResultsTypeString {
   return error
}
```

Let's keep the existing test case, and add another one where type is "" that is valid.

Reply:
do we call setDefaults on the result type? One option would be to add the default string type before validation

Reply:
> do we call setDefaults on the result type? One option would be to add the default string type before validation

Yes we call setDefaults on the result type and adds `string` as default type. But for existing resources as @afrittoli mentioned it doesn't work.

I will fix it soon!

Reply:
hmm, one way to deal with this is ensure that setDefaults is called even for the existing types. Is there a way we can ensure setDefaults is called for the existing types before we call validation?

Reply:
yup - something like https://github.com/tektoncd/pipeline/blob/067cbb7b71b2d77d5ac6166767b31ee6525447b7/pkg/reconciler/taskrun/taskrun.go#L120-L126

or https://github.com/tektoncd/pipeline/blob/8611c55275955ff4f8a88f57dfd2f0ecf8562470/pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go#L640-L644

Reply:
So the question is where should we call it to ensure for the existing resources. 🤔 @afrittoli Does this solution sounds good to you? We still can go with current solution, this may be a quick fix.